### ISSUE TITLE
azure pipelines builds are failing on Windows due to OpenSSH integration tests

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#9760]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#9760|
|Type|release blocker: release process bug|
|Created|2020-01-29 07:56:11Z|
|Branch|https://github.com/twisted/twisted/tree/9760-windows-build-failures|

for example, https://twistedmatrix.visualstudio.com/twisted/_build/results?buildId=1409&view=logs&j=4257c0a7-7659-5375-d6e0-cde9f28a3d36&t=46bdcfa9-535f-58a2-2f02-aba63a24289e&l=18475

this is a release blocker because tests are failing on every PR right now

<details><summary>Searchable metadata</summary>

```
trac-id__9760 9760
type__release_blocker__release_process_bug release blocker: release process bug
reporter__glyph glyph
priority__highest highest
milestone__None None
branch__9760_windows_build_failures 9760-windows-build-failures
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__None None
time__1580284571884434 1580284571884434
changetime__1582019175463886 1582019175463886
version__None None
owner__glyph glyph

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

I set up a local Windows development environment and can't reproduce these.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

After tweaking the tests to be just the tiniest bit verbose, I see this:

> 
Failure: twisted.conch.error.ConchError: ('exit code was not 0: 255 (b\'@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@\\r\\n@         WARNING: UNPROTECTED PRIVATE KEY FILE!          @\\r\\n@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@\\r\\nPermissions for \\\'dsa_test\\\' are too open.\\r\\nIt is required that your private key files are NOT accessible by others.\\r\\nThis private key will be ignored.\\r\\nLoad key "dsa_test": bad permissions\\r\\ntestuser@...: Permission denied (publickey).\\r\\n\')', None)


which clearly suggests the problem is that something about the permissions we're explicitly setting in the tests is getting ignored in this configuration.

This Stack Overflow question https://superuser.com/questions/1296024/windows-ssh-permissions-for-private-key-are-too-open indicates that this might be happening because our tmpdir now inherits undesirable permissions.

**Comment by twisted-trac** [comment]
|[<img alt="cjwatson's avatar" src="https://avatars.githubusercontent.com/u/2429141?s=50" width="50" height="50">](https://github.com/cjwatson)<a name="note_3"></a>|@cjwatson commented|
|-|-|

If it helps, I think this is the implementation of the relevant security check on Windows:

  https://github.com/PowerShell/openssh-portable/blob/latestw_all/contrib/win32/win32compat/w32-sshfileperm.c

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

Further investigation has yielded that the problem user is  S-1-5-32-545, which is `BUILTIN_USERS` according to https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/81d92bba-d22b-4a8c-908a-554ab29148ab

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

OK, I think this gets across the desired semantics for the test skip; hopefully CI will be happy with the result.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

The PR is here: https://github.com/twisted/twisted/pull/1220

**Comment by twisted-trac** [comment]
|[<img alt="cdunklau's avatar" src="https://avatars.githubusercontent.com/u/882547?s=50" width="50" height="50">](https://github.com/cdunklau)|@cdunklau set owner to @cdunklau|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="cdunklau's avatar" src="https://avatars.githubusercontent.com/u/882547?s=50" width="50" height="50">](https://github.com/cdunklau)<a name="note_9"></a>|@cdunklau set owner to @glyph|
|-|-|

I'm vaguely dissatisfied with the conditional skip in `ConchServerSetupMixin._createFiles`, but not enough to dig in my feet. I suggest a new ticket asking for help verifying that the tests that wind up calling that method on various platforms and versions.

LGTM, please merge

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

That's the idea with the "for real" followup I already filed / linked in the skip :). I am also not happy with this, but also; I really want to unblock the 20 other things that are backed up behind this.

Thanks for the review!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph set status to `closed`|
|-|-|

In changeset b719119403e3100ebc24d8e62ddbc0fa4c301ec0
```
#!CommitTicketReference repository="" revision="b719119403e3100ebc24d8e62ddbc0fa4c301ec0"
Merge pull request #1220 from twisted/9760-windows-build-failures

Author: glyph

Reviewer: cdunklau

Fixes: ticket:9760

Get continuous integration working again by fixing up some minor issues related to filesystem permissions and path configuration in tests.
```
